x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/PointInTimeSearchHelper.java

```diff
@@ -0,0 +1,126 @@
+            new OpenPointInTimeRequest(searchRequest.indices(), searchRequest.indicesOptions(), KEEP_ALIVE, null, null);
+
+        logger.trace("opening pit for indices: [{}]", Strings.arrayToCommaDelimitedString(searchRequest.indices()));
+        executeAsyncWithOrigin(client, SECURITY_ORIGIN, OpenPointInTimeAction.INSTANCE, openPointInTimeRequest,
+            ActionListener.wrap(openPointInTimeResponse -> {
+                final String pitId = openPointInTimeResponse.getSearchContextId();
+                final OriginSettingClient securityOriginClient = new OriginSettingClient(client, SECURITY_ORIGIN);
+                getNextPage(new ArrayList<T>(), pitId, null, securityOriginClient, searchRequest, hitParser, listener);
+            }, listener::onFailure));
+    }
+
+    private static <T> void getNextPage(List<T> results, String pitId, Object[] sortValues,
+                                        Client client,
+                                        SearchRequest searchRequest,
+                                        Function<SearchHit, T> hitParser,
+                                        final ActionListener<Collection<T>> listener) {
+        assert client instanceof OriginSettingClient : "client should have a specific origin";
+        final PointInTimeBuilder pointInTimeBuilder = new PointInTimeBuilder(pitId).setKeepAlive(KEEP_ALIVE);
+        if (results.isEmpty()) {  // first page
+            assert sortValues == null : "search_after should be null for the first page";
+            logger.trace("getting first page with pit: [{}]", pitId);
+            searchRequest.source().pointInTimeBuilder(pointInTimeBuilder);
+        } else {
+            logger.trace("getting next page with pit: [{}] and sort values: [{}]",
+                pitId, Strings.arrayToCommaDelimitedString(sortValues));
+            searchRequest.source().searchAfter(sortValues).pointInTimeBuilder(pointInTimeBuilder);
+        }
+        client.threadPool().executor(ThreadPool.Names.SAME).execute(() -> {
+            client.execute(SearchAction.INSTANCE, searchRequest,
+                ActionListener.wrap(searchResponse -> {
+                    processSearchResponse(searchResponse, results, client, searchRequest, hitParser, listener);
+                }, listener::onFailure));
+        });
+    }
+
+    private static <T> void processSearchResponse(SearchResponse searchResponse, List<T> results,
+                                                  Client client, SearchRequest searchRequest,
+                                                  Function<SearchHit, T> hitParser,
+                                                  final ActionListener<Collection<T>> listener) {
+        assert client instanceof OriginSettingClient : "client should have a specific origin";
+        final SearchHit[] hits = searchResponse.getHits().getHits();
+        results.addAll(Arrays.stream(hits).map(hitParser).collect(Collectors.toUnmodifiableList()));
```

Comment:
This requires creating an intermediate list for no reason. `toUnmodifiableList` isn't especially efficient.

```suggestion
        Arrays.stream(hits).map(hitParser).forEach(results:add);
```